### PR TITLE
generate/retrieve API key for user endpoint

### PR DESCRIPTION
**Pull request by flotoria**

### CODE REVIEW

File: api/main_endpoints/routes/User.js [hunk 1]
```diff
@@ -525,4 +526,39 @@ router.post('/usersValidVerifiedAndSubscribed', function(req, res) {
     });
 });
 
+// Generate an API key for the Messages API if the user does not have an API key; otherwise, return the existing API key
+router.post(('/retrieveAPIKey'), async (req, res) => {
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js]: i wanna use this endpoint from the command line, lets make it simple
```suggestion
router.post(('/apikey'), async (req, res) => {
```

File: api/main_endpoints/routes/User.js [hunk 2]
```diff
@@ -525,4 +526,39 @@ router.post('/usersValidVerifiedAndSubscribed', function(req, res) {
...
+    return res.sendStatus(FORBIDDEN);
+  } else if (!checkIfTokenValid(req)) {
+    return res.sendStatus(UNAUTHORIZED);
+  } else {
+    let { _id } = decodeToken(req);
+
+    User.findOne({_id})
+      .then((user) => {
+        if (!user) {
+          return res.sendStatus(UNAUTHORIZED);
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js]:
```suggestion
          return res.sendStatus(NOT_FOUND);
```

File: api/main_endpoints/routes/User.js [hunk 3]
```diff
@@ -525,4 +526,39 @@ router.post('/usersValidVerifiedAndSubscribed', function(req, res) {
     });
 });
 
+// Generate an API key for the Messages API if the user does not have an API key; otherwise, return the existing API key
+router.post(('/retrieveAPIKey'), async (req, res) => {
+  if (!checkIfTokenSent(req)) {
+    return res.sendStatus(FORBIDDEN);
+  } else if (!checkIfTokenValid(req)) {
+    return res.sendStatus(UNAUTHORIZED);
+  } else {
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js]: we dont need an else block if the first 2 `if`s are both returning

**Review by evanugarte — COMMENTED** [review]
wanna try adding tests? see the below file as an example

https://github.com/SCE-Development/Clark/blob/e2a203448f783ffdbdf581c8b8fd5dd4c6ec5f63/test/api/User.js#L496

File: api/main_endpoints/routes/User.js (hunk 3 above)
Comment (jonathanguven) [reply api/main_endpoints/routes/User.js]: word

**Review by jonathanguven — APPROVED** [review]
lgtm miweng

File: api/main_endpoints/routes/User.js [hunk 4]
```diff
@@ -525,4 +526,42 @@ router.post('/usersValidVerifiedAndSubscribed', function(req, res) {
...
 
+// Generate an API key for the Messages API if the user does not have an API key; otherwise, return the existing API key
+router.post('/apikey', async (req, res) => {
+  if (!checkIfTokenSent(req)) {
+    return res.sendStatus(FORBIDDEN);
+  }
+  // console.log("HELP!!!!!!!! ME!!!!!!!!", {
+  //   'req.body': req.body,
+  //   'checkIfTokenValid(req)': checkIfTokenValid(req),
+  // })
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js]:
```suggestion
```

**Comment by flotoria** [comment]
please merge this!

File: api/main_endpoints/routes/User.js [hunk 5]
```diff
@@ -35,6 +35,8 @@ const discordRedirectUri = process.env.DISCORD_REDIRECT_URI ||
 const logger = require('../../util/logger');
 
 const {sendUnsubscribeEmail} = require('../util/emailHelpers');
+const crypto = require('crypto');
+const { decode } = require('punycode');
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js]:
```suggestion
```

**Review by pnavab — APPROVED** [review]
b

File: api/main_endpoints/routes/User.js [hunk 6]
```diff
@@ -525,4 +526,38 @@ router.post('/usersValidVerifiedAndSubscribed', function(req, res) {
...
+          User.updateOne({_id}, {apiKey})
+            .then((result) => {
+              if (result.n == 0) {
+                return res.sendStatus(UNAUTHORIZED);
+              }
+              return res.status(OK).send({apiKey});
+            });
+        } else {
+          return res.status(OK).send({apiKey: user.apiKey});
+        }
```
Comment (evanugarte) [inline api/main_endpoints/routes/User.js]:
```suggestion
      }
      // logic to generate api key or return existing api key
      if (user.apiKey) {
        return res.status(OK).send({apiKey: user.apiKey});
      }
      let apiKey = crypto.randomUUID();
      User.updateOne({_id}, {apiKey})
        .then((result) => {
          if (result.n == 0) {
            return res.sendStatus(UNAUTHORIZED);
          }
          return res.status(OK).send({apiKey});
        });
```

**Review by evanugarte — APPROVED** [review]

**Pull request merged**